### PR TITLE
Added the spi driver for stm

### PR DESCRIPTION
**Pull request by g-lesssard**

### CODE REVIEW

File: src/bsp/src/esp/include/SpiHeader.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    uint8_t connected : 1;
+    uint8_t hasNeighborData : 1;
+    uint8_t failedCrc : 1;
+    uint8_t unused : 5;
+};
+
+struct StmSystemState {
+    uint8_t hasHost : 1;
+    uint8_t failedCrc : 1;
+    uint8_t unused : 6;
```
Comment (xgroleau) [inline src/bsp/src/esp/include/SpiHeader.h:17]: Same as other PR' why isn't a bool?

File: src/bsp/src/esp/include/SpiStm.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    } m_rxState;
+
+    spi_slave_transaction_t m_transaction;
+
+    struct Message {
+        // These buffers need to be word-alligned for DMA.
+        WORD_ALIGNED_ATTR std::array<uint8_t, STM_SPI_MAX_MESSAGE_LENGTH> m_data;
+        uint32_t m_sizeBytes;
+        bool m_sent;
+    } m_inboundMessage, m_outboundMessage;
```
Comment (xgroleau) [inline src/bsp/src/esp/include/SpiStm.h:42]: again, why protected and not out of the class definition?

File: src/bsp/src/esp/include/SpiStm.h [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    StmHeader::Header* m_inboundHeader;
+
+    void updateOutboundHeader();
+
+  private:
+    static void notifyMaster();
+    static void transactionCallback(void* instance, spi_slave_transaction_t* transaction);
+
+    StaticSemaphore_t m_semaphoreBuffer;
+    SemaphoreHandle_t m_semaphore;
```
Comment (xgroleau) [inline src/bsp/src/esp/include/SpiStm.h:54]: Can the ESP use the propolis freertos libs for stack tasks and mutexes?

File: src/bsp/src/esp/src/SpiStm.cpp [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        std::memcpy(m_outboundMessage.m_data.data(), buffer, length);
+        // Padding with 0 up to a word-alligned boundary
+        for (uint8_t i = 0; i < (length % 4); i++) {
+            m_outboundMessage.m_data[length] = 0;
+            length++;
+        }
+        // Appending CRC32
+        *(uint32_t*)(m_outboundMessage.m_data.data() + length) =
+            calculateCRC32_software(buffer, length);
+        ;
```
Comment (xgroleau) [inline src/bsp/src/esp/src/SpiStm.cpp:53]: remove ;

File: src/bsp/src/esp/src/SpiStm.cpp [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        m_outboundMessage.m_sizeBytes = length;
+        m_isBusy = true;
+        notifyMaster();
+        xSemaphoreGive(m_semaphore);
+        retVal = true;
+    } else {
+        m_logger.log(LogLevel::Debug, "Failed to acquire semaphore");
+    }
+
+    return retVal;
```
Comment (xgroleau) [inline src/bsp/src/esp/src/SpiStm.cpp:64]: return false here and true at 59

File: src/bsp/src/esp/src/SpiStm.cpp [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+        }
+        // Appending CRC32
+        *(uint32_t*)(m_outboundMessage.m_data.data() + length) =
+            calculateCRC32_software(buffer, length);
+        ;
+        length += CRC32_SIZE;
+        m_outboundMessage.m_sizeBytes = length;
+        m_isBusy = true;
+        notifyMaster();
+        xSemaphoreGive(m_semaphore);
```
Comment (xgroleau) [inline src/bsp/src/esp/src/SpiStm.cpp:58]: What is critical here? the notifyMaster?

File: src/main.cpp [hunk 7]
```diff
@@ -10,21 +10,21 @@ extern "C" {
...
-        logger.log(LogLevel::Info, "System has %d cores", info.m_cores);
-        if (info.m_osType == ChipInfo::ESP) {
-            logger.log(LogLevel::Info, "System is running on target");
+        ISpiStm* spiStm = &BspContainer::getSpiStm();
+        if (!spiStm->isBusy()) {
+            char message[] = "Hello STM";
+            spiStm->send((uint8_t*)message, sizeof(message));
         } else {
-            logger.log(LogLevel::Info, "System is running locally");
+            // logger.log(LogLevel::Info, "Spi driver is busy...");
```
Comment (xgroleau) [inline src/main.cpp:25]: remove comment?

File: src/taskManager/src/AbstractTask.cpp [hunk 8]
```diff
@@ -20,6 +20,8 @@ template <unsigned int stackSize>
 void AbstractTask<stackSize>::start() {
     m_taskHandle = xTaskCreateStatic(wrapper, m_taskName, stackSize, this, m_priority,
                                      m_stackArray.data(), &m_taskBuffer);
+    m_taskHandle = xTaskCreateStatic(wrapper, m_taskName, stackSize, this, m_priority,
```
Comment (xgroleau) [inline src/taskManager/src/AbstractTask.cpp:23]: Again, does the propolis lib works here?

**Review by xgroleau — CHANGES_REQUESTED** [review]
GJ! I didn't duplicate the comments on hivemind, but a lot of them are applicable here too for the state machine for example.

File: src/bsp/src/esp/src/SpiStm.cpp (hunk 6 above)
Comment (g-lesssard) [reply src/bsp/src/esp/src/SpiStm.cpp]: Removed semaphore, proved to be useless

File: src/taskManager/src/AbstractTask.cpp (hunk 8 above)
Comment (g-lesssard) [reply src/taskManager/src/AbstractTask.cpp:23]: Changed to propolis

File: src/bsp/src/esp/src/SpiStm.cpp [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+        updateOutboundHeader();
+        m_transaction.tx_buffer = &m_outboundHeader;
+        txLengthBytes = StmHeader::sizeBytes;
+        break;
+    case transmitState::SENDING_PAYLOAD:
+        m_transaction.tx_buffer = m_outboundMessage.m_data.data();
+        txLengthBytes = m_outboundMessage.m_sizeBytes;
+        break;
+    case transmitState::ERROR:
+        break;
```
Comment (xgroleau) [inline src/bsp/src/esp/src/SpiStm.cpp:144]: Log if there is an error, unless the error is frequent, but I don't think it's the case here

File: src/logger/src/Logger.cpp [hunk 10]
```diff
@@ -1,54 +1,24 @@
...
-        va_list args;
-        m_ui.printError("Error: Logger semaphore could not be created", args);
-    }
-
-    xSemaphoreGive(m_semaphore);
 }
 
-Logger::~Logger() { vSemaphoreDelete(m_semaphore); }
+LogRet Logger::log(LogLevel level, const char* format, ...) {
+    LockGuard lock = LockGuard(m_mutex);
```
Comment (xgroleau) [inline src/logger/src/Logger.cpp:11]: Create the lock guard in the if at line 13 , you don't need it if  the levels don't match

File: src/bsp/src/esp/src/SpiStm.cpp [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+        return false; // Not available
+    } else if (length >= STM_SPI_MAX_MESSAGE_LENGTH) { // Message too long
+        m_logger.log(LogLevel::Error,
+                     "StmEsp: Message length of %d is larger than maximum allowed of %d", length,
+                     STM_SPI_MAX_MESSAGE_LENGTH);
+        return false;
+    } else {
+        m_logger.log(LogLevel::Debug, "Sending message of length %d", length);
+        // Memcpy necessary to have buffer word-alligned for transfer
+        std::memcpy(m_outboundMessage.m_data.data(), buffer, length);
```
Comment (xgroleau) [inline src/bsp/src/esp/src/SpiStm.cpp:50]: Could you just link the doc here, like this link
https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/spi_slave.html#restrictions-and-known-issues

**Review by xgroleau — APPROVED** [review]
Very minor comment so I approved, allowing you to merge right away when you fixed them

File: src/logger/src/Logger.cpp (hunk 10 above)
Comment (g-lesssard) [reply src/logger/src/Logger.cpp:11]: I will also make the change in HiveMind, as it is the exact same code

**Pull request merged**